Add a Pulumi 2.0 Update blog post (#2599)
pulumi/docs: content/blog/pulumi-2-0-update/index.md

---
title: "An Update on Pulumi 2.0"
authors: ["joe-duffy"]
tags: ["Pulumi-News"]
date: "2020-03-06"
meta_desc: "Learn about some of Pulumi's progress rounding out our .NET and Go SDKs in addition to infrastructure validation using testing and Policy as Code."

> meta_image: "pulumi-1-0.png"

Reviewer: Nit: Are you planning to use a different meta image? Not sure why this is using a file named "1-0" for an update on 2.0.

Author: I'm just reusing the same header preview image from the 1.0 launch since we don't have the new 2.0 copy.

We've overhauled [our Go SDK ]({{< relref "/docs/intro/languages/go" >}})to be more idiomatic and easier to use. The changes here are a major improvment as you can see [in this before/after PR](https://github.com/pulumi/examples/commit/b88f385a8951f81fe41cb4f40eb940b074e9263b#diff-07803f7bc3bff97871e6e3eea7287cb8). We previously used `map[string]interface{}` types in many places instead of properly-typed data structures. Not any more &mdash; now you get strong typing for all Go resource APIs across all providers, including AWS, Azure, GCP, and Kubernetes. Examples and documentation have been converted to this new style. [This blog post]({{< relref "/blog/go-sdk-road-to-2" >}}) provides an overview of this new programming model approach.

Although we've had GoDoc support for some time now, we never linked to them! [Now we do.]({{< relref "/docs/reference/pkg?language=golang ">}}) We've also added Go versions to our getting started guides for [AWS]({{< relref "/docs/get-started/aws?language=golang" >}}), [Azure]({{< relref "/docs/get-started/azure?language=golang" >}}), and [GCP]({{< relref "/docs/get-started/gcp?language=golang" >}}), and have many more improvements on the way.

> There are a few core features still remaining to flesh out. The biggest one is retiring Dep in favor of Go modules everywhere. We are also adding support for aliases, transformations, and improved API projections, such as supporting union types. Finally, we'll be adding Go support for our Kubernetes, Docker, and Terraform remote state packages.

Reviewer 2: `aliases and transformations` -- this is already shipped and documented. 

`supporting union types` -- Is this true? I'm not aware of any plans for this.

Author: Great, I started this a while back - you all are moving fast!

Isn't https://github.com/pulumi/pulumi/issues/3803 about union types? It seems to be marked 2.0, although I'll admit it's light on details so I might be misunderstanding.

Reviewer 2: @lukehoban @pgavlin thoughts here? I'm not familiar with this particular item.

## Policy as Code

> [CrossGuard, our Policy as Code framework]({{< relref "/docs/guides/crossguard" >}}), is now ready to take for a serious test drive. This feature enables you to write policies in real code and enforce them during updates. Policies can check for anything, however common checks include security, compliance, cost management, and general and team best practices. You can write your own policy packages or use off-the-shelf policy packs like our own AWSGuard package.

Reviewer: Took Sophia's suggestion and also added a link around AWSGuard.

```suggestion
[CrossGuard, our Policy as Code framework]({{< relref "/docs/guides/crossguard" >}}), is now ready to take for a rigorous test drive. This feature enables you to write policies in real code and enforce them during updates. Policies can check for anything; however, common checks include security, compliance, cost management, and general and team best practices. You can write your policy packages or use off-the-shelf policy packs like our [AWSGuard]({{< relref "/docs/guides/crossguard/awsguard" >}}) package.
```

CrossGuard is open source and the functionality is behind [the new `pulumi policy` command]({{< relref "/docs/reference/cli/pulumi_policy" >}}), as well as [the new `--policy-pack` flag]({{< relref "/docs/reference/cli/pulumi_up#options" >}}) for the `preview` and `up` commands. You can use this on any edition of Pulumi, without restriction, including the offline backends.

> If you choose to use CrossGuard with the Enterprise Edition of Pulumi, however, you'll get some added functionality. This includes server-side enforcement of policies and organizational policies, including "policy groups" which let you group and apply many policies together to your stacks, such as for instance applying different policies to production than development environments.

Reviewer 3: ```suggestion
If you choose to use CrossGuard with the Enterprise Edition of Pulumi, however, you'll get some added functionality. This includes server-side enforcement of policies and organizational policies, including "policy groups" which let you group and apply many policies together to your stacks, such as applying different policies to production than development environments.
```

> If you choose to use CrossGuard with the Enterprise Edition of Pulumi, however, you'll get some added functionality. This includes server-side enforcement of policies and organizational policies, including "policy groups" which let you group and apply many policies together to your stacks, such as for instance applying different policies to production than development environments.

Reviewer 4: Is it really a "server-side" enforcement? I through policies are still evaluated inside `pulumi up`?

Author: The terminology we use is that the server _enforces_ the rules that get run, although the client _evaluates_ them.

![Policy Enforcement in the Console](./PolicyEnforcement.png)

We recently added the ability to tag policy packs with a semantic version so that you can easily manage and apply packs using user-friendly versions instead of auto-generated numbers.

The remaining areas of focus between now and launch include:

> * Enabling policy pack authoring in Python. Policies can already apply to stacks written in any language, no matter the language of the policy pack itself, but today policy packs are authored in Node.js. Python is coming soon and Go and .NET will follow afterward.

Reviewer 3: ```suggestion
* Enabling policy pack authoring in Python. Policies can already apply to stacks written in any language, no matter the language of the policy pack itself, but today policy packs are authored in Node.js. Python is coming soon, with Go and .NET following afterward.
```

> * Allowing policy packs to be configured. This will let you selectively enable or disable specific rules, and/or set configuration variables on them, that influence their behavior, on a per stack or policy group basis.

Reviewer 3: ```suggestion
* Allowing policy packs to be configured. Configuring policy packs lets you selectively enable or disable specific rules, and/or set configuration variables on them, that influence their behavior, on a per stack or policy group basis.
```

> * Reference policy packs for AWS, Azure, GCP, and Kubernetes. These packs will be usable out-of-the-box with a collection of some of the most important configurable rules we've seen, in addition to being a starting point for creating your own packs.

Reviewer 3: ```suggestion
* Making reference policy packs available for AWS, Azure, GCP, and Kubernetes. These packs will be usable out-of-the-box with a collection of some of the most important configurable rules we've seen, in addition to being a starting point for creating your packs.
```

> [CrossGuard is ready to try today]({{< relref "/docs/get-started/crossguard" >}}) and [we've begun blogging about some example use cases]({{< relref "/blog/manage-infrastructure-with-pac" >}}) that you may find interesting.

Reviewer 3: ```suggestion
[CrossGuard is ready to try today]({{< relref "/docs/get-started/crossguard" >}}), and [we've begun blogging about some example use cases]({{< relref "/blog/manage-infrastructure-with-pac" >}}) that you may find interesting.
```

> ## Other Goodies

Reviewer 5: Do you also want to mention YAML export for k8s?

Author: Good idea! Will add.

In addition to the major areas above, we've continued to ship many other improvements:

> * [Stack export for the CLI and in the service]({{< relref "/docs/reference/cli/pulumi_stack_export" >}}), enabling you to export any past version of a stack's state file. This is helpful for debugging, restoring or rolling back changes, or even just for historical reporting purposes.

Reviewer 3: ```suggestion
* [Stack export for the CLI and in the service]({{< relref "/docs/reference/cli/pulumi_stack_export" >}}), enabling you to export any past version of a stack's state file. This helps you debug, restore, or roll back changes, or even just for historical reporting purposes.
```

* [Many other quality of life improvements in the Pulumi SaaS Console]({{< relref "/blog/pulumi-service-improvements_02-2020" >}}), including stack tagging, grouping, and sorting; deep linking into CI/CD systems; infrastructure configuration pretty-printing; reverse stack membership lookups; performance improvements; and more.

> * [Full audit logs for Pulumi Enterprise Edition]({{< relref "/blog/auditing-your-organizations-infrastructure-as-code-activity" >}}), enabling you to go back in time and figure out who, when, and why something about your organization, project, or stack changed.

Reviewer 3: ```suggestion
* [Full audit logs for Pulumi Enterprise Edition]({{< relref "/blog/auditing-your-organizations-infrastructure-as-code-activity" >}}) enables you to go back in time and figure out who, when, and why something about your organization, project, or stack changed.
```

> * [Pulumi runtime mocking for infrastructure testing](https://github.com/pulumi/pulumi/pull/3738), developed in partnership with several of our customers leading the charge on "test driven infrastructure." Look for more end-to-end docs and guidance soon!

Reviewer 3: ```suggestion
* [Pulumi runtime mocking for infrastructure testing](https://github.com/pulumi/pulumi/pull/3738), developed in partnership with several of our customers leading the charge on "test-driven infrastructure." Look for more end-to-end docs and guidance soon!
```

> * [We've been slowly adding more and more providers and now have over 30 of them]({{< relref "/docs/reference/pkg" >}}), including Aiven, Cloudflare, Consul, Datadog, DigitalOcean, DNSimple, F5 BIG-IP, Fastly, GitLab, Kafka, Linode, MySQL, New Relic, Okta, Packet, PostgreSQL, RabbitMQ, SignalFX, Vault, and more.

Reviewer 3: ```suggestion
* [We've been adding more and more providers and now have over 30 of them]({{< relref "/docs/reference/pkg" >}}), including Aiven, Cloudflare, Consul, Datadog, DigitalOcean, DNSimple, F5 BIG-IP, Fastly, GitLab, Kafka, Linode, MySQL, New Relic, Okta, Packet, PostgreSQL, RabbitMQ, SignalFX, Vault, and more.
```

## Goodies Still to Come

> Our API docs are getting a major overhaul! This includes a more resource-oriented view which will make it much easier to navigate and find what you're looking for, including code samples. This will also bring all languages on equal footing in terms of how documentation is presented and coverage of code samples. Expect to see some of this land and evolve in the weeks to come.

Reviewer 3: ```suggestion
Our API docs are getting a major overhaul! The API docs us a more resource-oriented view, which will make it much easier to navigate and find what you're looking for, including code samples. This also makes all languages more uniform in terms of how documentation is presented, and coverage of code samples.  Expect to see some of this land and evolve in the weeks to come.
```

> We continue to make progress on our multi-language library sharing capabilities. These will enable us to bring your favorite packages like EKS and AWSX to new languages, including Python, C#, and Go, in addition to letting you write your own. This has entailed some gnarly engineering work but will [give you access to our EKS in Python](https://github.com/pulumi/pulumi-eks) to start.

Reviewer 3: ```suggestion
We continue to make progress on our multi-language library sharing capabilities. These will enable us to bring your favorite packages like EKS and AWSX to new languages, including Python, C#, and Go, in addition to letting you write your own. Expanding our multi-language library has entailed some inspired engineering work but will [give you access to EKS in Python](https://github.com/pulumi/pulumi-eks) to start.
```

> Our [`tf2pulumi` tool](https://github.com/pulumi/tf2pulumi), which converts any Terraform HCL to Pulumi code, is currently getting a fresh coat of paint. This includes HCL2 support and Python code-generation.

Reviewer 3: ```suggestion
Our [`tf2pulumi` tool](https://github.com/pulumi/tf2pulumi), which converts any Terraform HCL to Pulumi code, is currently getting a fresh coat of paint. Look for HCL2 support and Python code-generation when released.
```
